app/lib/shared/cached_value.dart

```diff
@@ -0,0 +1,75 @@
+// Copyright (c) 2021, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:logging/logging.dart';
+import 'package:meta/meta.dart';
+import 'package:retry/retry.dart';
+
+final _logger = Logger('cached_value');
+
+typedef UpdateFn<T> = Future<T> Function();
+
+/// A locally cached and periodically updated value that can be accessed
+/// synchronously.
+class CachedValue<T> {
+  final UpdateFn<T> _updateFn;
+  final RetryOptions _retryOptions;
+  Timer _timer;
+  T _value;
+  Completer _ongoingCompleter;
+  bool _initialized = false;
+
+  CachedValue({
+    @required UpdateFn<T> fn,
```

Comment:
Do we need other options than `fn` maybe it doesn't need to be a keyword parameter?